### DISCUSSION TITLE
Feedback: New options for controlling the default commit message when merging a pull request

### DISCUSSION THREAD

**Discussion by willsmythe** [community/community#30439] [Pull Requests]
We recently shipped more options for customizing the default commit message presented when merging a pull request, including using the pull request's description! The goal is more predictable, consistent, and useful commit messages for your pull requests.

## How it works

From repository settings, a user with maintainer or admin access can choose the default format for commit messages produced when merging a pull request. 

<img width="790" alt="image" src="https://user-images.githubusercontent.com/2503052/185104685-6fccc041-2379-4c56-969c-6732d10821f6.png">

Learn more about [configuring merge commit merging](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/configuring-commit-merging-for-pull-requests) and [configuring squash merging](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/configuring-commit-squashing-for-pull-requests).

## Thoughts?

Let us know what you think and what would make the feature more useful.

**Comment by bewuethr** [comment]
This is great, no more test / test / fix typo / ... commit messages when squashing! The only drawback I can see for the "default to pull request title and description" message is that a paragraph from the description becomes a single, potentially very long line without editing:

This PR description

![image](https://user-images.githubusercontent.com/8521043/186280839-6d42eabd-dfb1-433f-8e7f-2c40ba5261d8.png)

becomes this log message:

![image](https://user-images.githubusercontent.com/8521043/186281327-db143308-b7ce-46e5-bd1e-11d9c273da04.png)

I know that the PR description field is a bit special in that it's Markdown, but linebreaks are respected, so what looks good in a log message looks weirdly short-lined in the PR description.

Is it thinkable to word-wrap the description at 80 characters when converting to log message? Might not be perfect, but better than super long lines, maybe?

**Reply by adamshapiro0** [reply]
Or better yet, maybe make this configurable at a project or organization level? And have the text wrapping properly handle bulleted lists and other markdown.

We generally prefer to have the commit details be a verbatim copy of what was in the PR description. Right now, it seems to be randomly wrapping bulleted lists -- some of them wrap after ~60 chars, others don't wrap at all. For the ones that do wrap, it is not indenting the new lines correctly to match the original indent.

For some descriptions, that won't make much difference, though 60 chars does seem unnecessarily small. For more involved descriptions, this wrapping can make things much harder to read. For now, we have to undo the wrapping by hand before merging.

**Reply by brasic** [reply]
@adamshapiro0 GitHub engineer here. Would you mind sharing some example PR description lines that didn’t wrap properly when translated to a commit message? Feel free to change the phrasing to nonsense if it’s private info. You can also email it to brasic@github.com.

**Reply by adamshapiro0** [reply]
@brasic sure thing. Just sent you an email with an example, though I also just made a test PR with the description that didn't work last week and it seems to be working now. Not sure if something got fixed in production?

**Reply by bewuethr** [reply]
Hey, this (my original issue) has been fixed! Most excellent. 🚀

**Reply by SanjayVas** [reply]
Is there any way to turn the wrapping behavior off?

**Reply by SanjayVas** [reply]
I know this feedback is very old, but I have another important reason for the wrapping to be off: the current wrapping logic breaks trailers. The [git trailers specification](https://git-scm.com/docs/git-interpret-trailers) requires that multi-line trailers be indented. This means that if what is being wrapped is the value of a trailer, it needs to be indented. Of course, it may be difficult to figure out if something is intended to be a trailer. It's much easier if line wrapping can just be turned off.

**Comment by kerasing** [comment]
What's the roapmap for making this part of GHE?

**Reply by willsmythe** [reply]
It's planned for the next release (3.7).

**Comment by lwhiteley** [comment]
Awesome and long awaited.


Can there also be an option for only the description?
The title can sometimes be a bit redundant when viewing the commit history

![Screenshot_20220824_101524](https://user-images.githubusercontent.com/6108793/186367535-9436dbe1-b0b8-4968-95b4-2c5a14284e01.png)

Thanks for these cool options!

**Comment by karfau** [comment]
Any plans to configure a default per organization?
We have plenty of repositories and it would be great to be able to configure the defaults when a new repo is created, or even apply those defaults to all repos that have been created before this setting was an option.

(Especially since the default is the old behavior)

**Comment by MPLew-is** [comment]
If you set the merge commit message to `Default to pull request title and description`, and the PR does not have a description, you can't even access the option to merge your PR and change the message until you either change the setting or add a description:

![Screen Shot 2022-08-26 at 17 26 12](https://user-images.githubusercontent.com/11891085/187006962-0689ea30-0b88-4abd-82eb-bde98f43f879.png)

**Reply by MPLew-is** [reply]
This is what the description for that PR looks like:
![Screen Shot 2022-08-26 at 18 30 59](https://user-images.githubusercontent.com/11891085/187007043-3b8f2430-9cf6-47d3-8b05-9fd158dab0cf.png)

**Reply by Klaitos** [reply]
Same issue here, if you put anything in the description it fix the issue for information.

**Reply by MPLew-is** [reply]
This appears to be fixed now.

**Reply by Mina012015** [reply]
> نفس المشكلة هنا ، إذا وضعت أي شيء في الوصف ، فسيتم إصلاح المشكلة للحصول على المعلومات.

ع

**Comment by tiit-clarifai** [comment]
Our PR template contains markdown comments to provide instructions on what to write in the description. These are not visible in the rendered view so are often left intact by developers, but are undesirable noise in the plaintext commit message.

Would you consider adding an option to omit comments from the PR description when used as the default commit message?

For example, let's say you have the following PR description:
```
<!-- Please include a summary of the changes proposed in this pull request. -->
This PR updates our Petshop API client to version 2.0.

<!-- If there is a related issue to this PR, then reference it here. -->
Fixes #1234

<!-- ### Notes -->
<!-- If you have any additional notes, then uncomment the section above and write them here. -->
```
Enabling the option would generate the following commit message when used with **Pull request title and description**:
```
Pull Request Title (#5678)

This PR updates our Petshop API client to version 2.0.

Fixes #12345
```

**Reply by brasic** [reply]
I think perhaps rather than an option it makes more sense to always omit HTML comments. Can you suggest a scenario where these would be wanted in the commit message?

**Reply by tiit-clarifai** [reply]
Not really, I am fine with always omitting.

**Reply by rmacklin** [reply]
If a commit message includes a code snippet which contains an HTML comment, ideally the code snippet would not be modified. (In other words, I think we would only want to omit HTML comments that are _not_ part of a code snippet.)

**Reply by willsmythe** [reply]
Good suggestion. We have an issue tracking this internally.

**Reply by kriskowal** [reply]
Where can the public track progress? Thank you, omitting guidance comments would be very fine.

**Reply by ghost** [reply]
@willsmythe any update on this? at the time of writing, there are 44 up-votes.

**Reply by theprivileges** [reply]
This is a feature that would save us a lot of time.  Right now we have to manually remove these HTML comments from our template before merging with **pull request title and description**.

**Reply by Alex-ley-scrub** [reply]
+1 for this

**Comment by jplatte** [comment]
It would be really nice to be able to just set a template. The dropdown options right now really aren't self-explanatory. None of them suggest the PR number will be included, which I expected from the blog post title I could finally get rid of…

**Reply by willsmythe** [reply]
Templates is something we want to support in the future, just didn't make the cut for this initial wave of improvements.

**Reply by zane131** [reply]
It's almost been two years. Any outlook on this?

**Comment by brandonbirdj** [comment]
When selecting option to use `pull request title` it appears to take the original title from when the pull request was created. I would expect it to take the PR title at the time of merging. 

For example many actions such as https://github.com/marketplace/actions/pull-request-ticket-check-action automatically update the PR title.

**Reply by willsmythe** [reply]
Did you switch the new repo-level setting to something other than the default commit message format? The default commit message format for squash merge is the commit message (if the PR contains only 1 commit), which is sometimes the same as the PR title, but obviously can be different.

**Reply by brandonbirdj** [reply]
I've updated to clarify that this finding is for when the new setting of using `pull request title` is used.

**Comment by EricStG** [comment]
Is it possible to configure this via `.github/settings.yml`? I can't find any doc for it

**Reply by bewuethr** [reply]
Isn't that file provided/read by a [third-party app](https://github.com/apps/settings)?

**Reply by EricStG** [reply]
It is, my bad!

**Comment by bewuethr** [comment]
tl;dr: It would be nice to be able to select defaults based on PR author.

---

My use case is pretty much to optimize for the case where the merger doesn't edit the merge commit message at all. Using "Default to pull request title and description" gets us pretty far for that, and it can be improved with linters running on the PR description and similar.

However, one case that isn't covered and leads to an ugly default is dependabot PRs. The description on a dependabot PR is very, very long, and makes for a terrible commit message.

I'm not sure what a great default could be. Maybe the ability to set the behaviour depending on the PR author? If opened by dependabot, then use "Default message", which for dependabot PRs would be the very concise "Bump X from vA.B.C to vX.Y.Z" and some details in the description:

![image](https://user-images.githubusercontent.com/8521043/193926894-8b938b59-88b8-4a89-8279-2ffc246d7f8c.png)

And for all other authors use "Default to pull request title and description".

<details><summary><i>Click to see (not even all of) a commit with the PR description from dependabot as the message</i></summary>

![image](https://user-images.githubusercontent.com/8521043/193926169-1ae4e06b-df34-4175-8bd4-4d5aa1ca39bf.png)

</details>

**Reply by Alex-ley-scrub** [reply]
+1 for this - it is super annoying at the minute

**Comment by luisrayas3** [comment]
I would like to request a way to specify boundaries for what to include in the commit message. The use case for our team is that our template includes a checklist of things related to the PR that don't make sense or would be too noisy to be included in the git history. In our case a comment for which everything _below_ is excluded would be sufficient but both begin/end might make sense for completeness. Example:

```markdown
<!-- github-commit-message-begin -->  <!-- This line would ideally be optional -->
This is the main description
which we do want in the commit message body.

<!-- github-commit-message-end -->
## Merge checklist
- [ ] This is too much
- [ ] We don't want it in the commit message
```

**Comment by igorsafo** [comment]
In case of a project that has multiple repositories with different set of branches, PRs and issues, adding a PR number into the commit title by default results in weird references to wrong PRs when the branches between repositories are synced. It would be nice to have an option to remove a PR number from the title:
- repo 1: commit "foo (#PRX)" -- points to a PR X in repo 1
- repo 2: commit "foo (#PRX)" -- points to a PR X in repo 2 that is different from the original one

**Comment by daniellowell** [comment]
Can we have an option that **does not** include the PR title? In our org there is typically internal information that we do not want to be carried into the commit messages. As a safety it would be good to not have PR titles into commit messages as an option for all contributors.

**Comment by sinuscosinustan** [comment]
What I'm currently missing is an option to add trailers such as `Merged-by`, `Reviewed-by` or `Approved-by` to the merge commit message.

This makes it easier to identify who reviewed or approved a PR by simply checking the git log.

**Comment by kaidokert** [comment]
Would it ever be possible to optionally _disable editing_ the default merge message altogether ? We are trying to keep a clean standard as to what goes into the merged commit message, and that's all doable with workflow runs - we can check the PR title, description, commits and anything else we need.

However, the final action of submitting/merging happens after all of the checks have passed, and it gives everyone who merges full control to submit anything they want.

**Reply by bewuethr** [reply]
You could get there with a branch protection that lets only a bot account merge, and people trigger a merge with a comment (similar to `@dependabot merge`) for the bot to then merge in the way you want.

**Reply by kaidokert** [reply]
Thanks for the idea - yes this might work as a crutch. It would make the user interface less than obvious though, and would need extra thought if we allow both squash and merge and regular merge submits. Perhaps merge queue would give us a similar option, i haven't really looked yet.

**Comment by kimhanse** [comment]
Could this feature be extended to also affect the commit message prepared when using auto-merge?

**Reply by karfau** [reply]
I don't understand your question, this is all that this feature is about.

**Comment by thomaseizinger** [comment]
At the moment, the PR number is included in brackets at the end of the title. However, the standard way of linked to things in commit messages is to use trailers. It would be great if there was an option to either configure that or change the default. I'd like to see something like:

```
This is my commit title

Here is a body. This used to be the PR description.

Pull-Request: #1234.
```

**Comment by BenSluis** [comment]
It would be super helpful if the default could be controlled not just at the repository level but also in specific branch protection rules.

**Comment by max-hartman** [comment]
It would also be great to be able to have the entire URL for the PR in the last line of the commit body (could ideally be set programmatically in a template. The PR number at the top is nice, but when I'm in my terminal and I want to look a commit, I don't have a way to click into and open the PR for that commit. Some IDEs try to provide support for this by having "Issue Navigation" (allowing regex matching on snippets of the commit to create a URL from), but since its at the project level, once I have multiple repos in the same workspace it doesn't know which repo PR (# 123) belongs to.

**Comment by schoetbi** [comment]
I second  @BenSluis that there also should be a relation to the rules that can be setup to ensure a certain format for commit messages. For example, we want to ensure with a rule that the commit message for a pull request merge message adheres to be in the format of a [conventional commit](https://www.conventionalcommits.org/en/v1.0.0/) message. We need this to be able to generate the release notes later automatically with [git cliff](https://github.com/orhun/git-cliff). 

So the workflow should be like this:
- The PR creator prepares the first line of the commit message in the PR title
- Details for the commit can be placed in the PR description
- If the target branch has any rules for the commit message, this should be shown as a prerequisite before the merge is allowed
- If any rule is not fulfilled, the rule name that avoids this should be shown in red.

**Comment by amartyaroy1** [comment]
hello @willsmythe 
I liked this feature which is rolled out.I am curious to know is it possible to get these settings via grapql also.I tried via rest api so there is a field named as `use_squash_pr_title_as_default`: true/false which signifies that these settings are enabled or not. But i want to check whether these new settings are enabled or not in repos via graphql so can i get it? Can you please point which field will give me the resultant output?

I tried out these fields 
 mergeCommitTitle: "MERGE_MESSAGE",
            squashMergeCommitMessage: "COMMIT_MESSAGES",
            squashMergeCommitTitle: "COMMIT_OR_PR_TITLE",
but these are the not the ones that we want

**Comment by Martiix** [comment]
This is great! However, I would also like to be able to set a default commit message format for the revert button in the GUI:
![image](https://github.com/community/community/assets/31621101/efb8211e-2c31-4a42-80ca-9221fd665a83)
Any chance this will happen anytime soon? :)

**Comment by nayefserag** [comment]
Is it possible to configure this via .github/settings.yml ?

**Comment by mark-wiemer** [comment]
Is this feature still supported? I can't find this setting anywhere, am I blind?

**Reply by bewuethr** [reply]
It's on the top-level settings page of any specific repository, like `https://github.com/OWNER/REPO/settings`, under "Pull requests":

![image](https://github.com/community/community/assets/8521043/1a628dbe-cb21-484a-8883-492e26fdaa53)

**Reply by mark-wiemer** [reply]
Thanks! I kept looking for it as an option in the settings nav or from the branch rules section.

**Comment by zane131** [comment]
This is *almost* exactly what I need. However, this feature adds `Signed-off-by:` and `Co-authored-by:` trailers to the end of the message and inserts `------` in between those trailers and the trailers in the PR description. 

![image](https://github.com/user-attachments/assets/93aec975-9c7c-4bb8-a208-de47b7221689)

While the autogenerated trailers are appropriate (and a nice feature add), the separation of the two sets of trailers breaks commands like `git log --format="%(trailers)"` because only the last block of text in the commit message is recognized as trailers and the important trailers from the PR description are effectively ignored.

Also, as stated [above](https://github.com/orgs/community/discussions/30439#discussioncomment-5171022), the PR number added to the title only has context to the original repository and causes confusion in forks.

**Comment by Takym** [comment]
Are there any options for using a custom default commit message?

**Comment by AndiDog** [comment]
Voting against putting the PR description into the commit message. Auto-adding the description is often just unnecessary bloat. Do you  want a commit message containing this? It's a real, rendered, filled-out example for a perfectly fine PR template of an open source project, but only relevant in the PR review, not in the commit message:

```
What type of PR is this?
/kind feature

What this PR does / why we need it: Implements the MachinePool Machines clusterAPI proposal

Which issue(s) this PR fixes (optional, in fixes #<issue number>(, fixes #<issue_number>, ...) format, will close the issue(s) when PR gets merged):
Fixes https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/4184

Special notes for your reviewer:

Checklist:

    squashed commits
    includes documentation
    adds unit tests
    adds or updates e2e tests

Release note:

Added AWSMachines to back the ec2 instances in AWSMachinePools and AWSManagedMachinePools.
```

The reason why I upvoted this feedback idea is because users typically just click the merge button, accepting the default commit message without editing, so the default must be minimalistic and reasonable. Only the PR title can be expected to be fine in most cases. Organizations want to avoid committing stuff like this, just because the PR author needed to make a few expected pushes to make CI green, or finish the feature:

```
🌱 Implement the actual thing (#12345)

* Implement the actual thing

* Address review comments

Signed-off-by: The Author author@example.com

* Address some more review comments

Signed-off-by: The Author author@example.com

* Address smoe more review comments

Signed-off-by: The Author author@example.com

* Address more review findings

---------

Signed-off-by: The Author author@example.com
```

**Suggestion:**

- Keep only PR title and number as default commit message
- Optional: take over `Co-authored-by`, `Signed-off-by` and other well-known annotations, but remove duplicates
- Make that the global default across GitHub, so organizations don't have to configure something to get rid of the above, horrible squash commit messages that happen because users (rightfully) accept the default and hit the merge button.
- Allow PR author to choose from other presets before clicking the merge button, such as PR title+description and whatever other choices get implemented. This offers the people who _care_ about commit messages – those who don't click the merge button immediately – to more easily apply the desired template if the default doesn't suit them.

**Reply by spydon** [reply]
There is a beta feature rolled out to some repositories where you can remove parts of the PR description, see here for example:
https://raw.githubusercontent.com/flame-engine/flame/refs/heads/main/.github/pull_request_template.md
Then you will only get what the user has written.

What would be even better is if there was a form on the PR page, like the new forms for issues, so that the user doesn't have to manually edit markdown when submitting a PR.

**Reply by zane131** [reply]
>Voting against putting the PR description into the commit message. Auto-adding the description is often just unnecessary bloat.

It really depends on the use case. My organization utilizes elements on the commit message, particularly in the trailers, for tracking purposes. This isn't for everyone, but is incredibly helpful for us. 

Also, my organization has a set of maintainers that will do the "Squash and merge", which very likely will not be the author of the pull request. By utilizing the description we know what the author intended to have as the commit message of the squashed pull request. This takes the burden off of the maintainers so that they do not need to generated a formatted commit message. Instead, they can, as you say, "just click the merge button."

Again, it all depends on the use case. That is why it is nice to have the option.

**Reply by luisrayas3** [reply]
@spydon can you provide any links or information about this beta feature? I can't find any information and I'm desperate for exactly this for my org.

**Reply by spydon** [reply]
@luisrayas3 It was @anleac that helped us set it up, I don't think it's a documented feature yet.

**Reply by Alex-ley-scrub** [reply]
Hi @anleac would you also be able to help me / us set this up for our Org please?

**Reply by anleac** [reply]
👋 apologies for the late reply - I've been on vacation, and unfortunately, I no longer work at GitHub to be able to progress with the roll-out of this feature. @willsmythe do you think you could look into the progressed roll out, or at least, flag-in @Alex-ley-scrub's org?

**Reply by luisrayas3** [reply]
@willsmythe this is the number 1 pain for us. I'd be happy to beta-test this feature!

**Reply by Alex-ley-scrub** [reply]
@willsmythe do you still work at GitHub?
We now have a GitHub action we (heavily with CoPilot) wrote to pull out Dependabot extra info into a PR comment so that the info is not lost but the PR merge commit is much cleaner. But the best option would still be some configurable template / rules to ignore content between comment blocks and/or between `<details></details>` etc.

**Comment by bartgol** [comment]
I would still very much like to customize the message via some yml file in the `.github` folder. E.g., I would like to be able to add info such as the labels that were attached, or the assignees/reviewers list, or the name of who effectively merges the PR. E.g., a `merge-commit-default.yaml` file like the following (some may not be possible, but for just the sake of showing the idea)

```yaml
headline: Merge pull request {pr.number} from branch {pr.head.ref}

body: |
  PR Title: {pr.title}
  PR Author: {pr.author}
  PR Labels: {pr.labels}
  PR Reviewers: {pr.reviewers}
  PR Approved by: {pr.reviewers_approved}
  PR Assignees: {pr.assignees}
  
  Merged by: {user}
```

This way, I can quickly parse my git history for merging of PRs that had certain labels, or that were reviewed by user X, or merged by user Y.

Edit (adding here, to avoid starting another thread): I would also like a `{pr.body}` and `{pr.body_header}` option. The former would include the whole body of the opening pr message, while the latter would grab all lines until the first line matching (exactly) the string "---". This allows to format your PR message with a short summary at the top, and messages/explanations for reviewers (which are not meaningful in a commit msg) after the "hrule" line.

**Reply by Takym** [reply]
I suggest the below placeholders and flag to avoid <https://github.com/orgs/community/discussions/142044>.

* `{pr.number.safe}` placeholder - a number with the fork owner and repository name.
    * Format: `<FORK_OWNER_NAME>/<FORK_REPO_NAME>#<NUMBER>`
* `{pr.number.unsafe}` placeholder - a number only. Same with `{pr.number}`. This is for the private repositories.
    * Format: `#<NUMBER>`
* A flag to determine converting commit messages from unsafe issue/pull request numbers to safe. I propose names are as follows:
    * ~~`use_safe_numbers`~~
    * ~~`convert_numbers`~~
    * ~~`use_safe_issue_pr_numbers`~~
    * ~~`convert_issue_pr_numbers`~~
    * `issue_pr_number_format` with enum values: `no_converted`, `safe`, `unsafe`, `optimized`
        * (`optimized` means `safe` for other repositories, `unsafe` for the current repository)
    * `number_format` - `issue_pr_` is redundant?

Edit: I dropped `use_safe_numbers`, `convert_numbers`, `use_safe_issue_pr_numbers`, and `convert_issue_pr_numbers` and added `number_format`.

**Comment by keikoro** [comment]
Is there any documentation available for the format of `Default Message`?

I'm seeing merge commits in organization repos which are formatted "Merge pull request #N from ORG/REF", i.e. the source of a PR is the branch name prefixed with the organization name but missing the repository name. ... Is this normal? To be expected? Or due to a(n unfortunate or misconfigured) default set by the organization that I'm unaware of?

IIRC the source branch of a merge commit is prefixed with the user name of the PR author when it's coming from a fork rather than a branch directly on the target repo. Which might make sense to differentiate between "external" and "internal" code sources. But adding _the org name_ in front of the branch name _without_ also adding the repo name is weird and confusing.

**Reply by keikoro** [reply]
The docs seem to not accurately describe the ref formatting used in default messages on org repos, see: https://github.com/github/docs/issues/38163

**Comment by jsoref** [comment]
@queenofcorgis is this discussion really still open? It's over three years old.
